### PR TITLE
Adding Trainer

### PR DESCRIPTION
**Pull request by captain-pool**
Adding Trainer for Knowledge Distillation.

Addresses #42, #45 , #46 , #47, #44

### CODE REVIEW

File: E3_Distill_ESRGAN/libs/models/teacher.py [hunk 1]
```diff
@@ -1,5 +1,12 @@
+""" Module to load Teacher Models from Teacher Directory """
 import os
 import sys
+from libs import settings
 # Fetching Generator from ESRGAN
```
Comment (srjoglekar246) [inline E3_Distill_ESRGAN/libs/models/teacher.py]: newline between imports and code.

File: E3_Distill_ESRGAN/libs/settings.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+""" Module Containing Settings and Training Statistics Model YAML Handler """
```
Comment (srjoglekar246) [inline E3_Distill_ESRGAN/libs/settings.py]: More readable docstring please :-)

File: E3_Distill_ESRGAN/libs/train.py [hunk 3]
```diff
@@ -1 +1,230 @@
...
+    """
+      Args:
+        teacher: Keras Model of pre-trained teacher generator.
+                 (Generator of ESRGAN)
+        discriminator: Keras Model of pre-trained teacher discriminator.
+                       (Discriminator of ESRGAN)
+        summary_writer: tf.summary.SummaryWriter object for writing
+                         summary for Tensorboard.
+        data_dir: Location of the stored dataset.
+        manual: Indicate if data_parameter contains Raw Files or TFRecords.
```
Comment (srjoglekar246) [inline E3_Distill_ESRGAN/libs/train.py]: nit: Keep 'similar'-types arguments together. So group together `data_dir` and `model_dir`. Also, `manual` doesn't seem the best way to describe what that parameter indicates. Maybe just `raw_data`. And also `data_parameter` -> `data_dir`?

File: E3_Distill_ESRGAN/libs/utils.py [hunk 4]
```diff
@@ -1,5 +1,46 @@
...
+# Loading utilities from ESRGAN
+sys.path.insert(
+    0,
+    os.path.abspath(
+        settings.Settings(student=True)["teacher_directory"]))
+
+from lib.utils import RelativisticAverageLoss
+
+
+def save_checkpoint(checkpoint, name, basepath="", student=False):
```
Comment (srjoglekar246) [inline E3_Distill_ESRGAN/libs/utils.py]: `student` -> `use_student_settings`.

same below.

File: E3_Distill_ESRGAN/libs/train.py (hunk 3 above)
Comment (captain-pool) [reply E3_Distill_ESRGAN/libs/train.py]: Any suggestions how to indicate if it is `data_dir` or `manual_dir`, without using `manual`?

**Comment by captain-pool** [comment]
> Have we added any README or documentation that references the paper we are using for this? And what changes we are making to the loss function, etc?

We have 2 options for Loss. We can choose any of these, depending on the output we are getting.
- Comparative Loss -MSE Loss (As Stated by the Paper).
- **[Preferred]** Adversarial Loss - Joint MSE and Relativistic Average Loss.

I've implemented both, we will be using each of them based on the output we get.

File: E3_Distill_ESRGAN/libs/train.py (hunk 3 above)
Comment (srjoglekar246) [reply E3_Distill_ESRGAN/libs/train.py]: Wouldn't it be either of the two? So you can check which is not-empty?

Comment (captain-pool) [reply E3_Distill_ESRGAN/libs/train.py]: okay! Cool. :)

Comment (captain-pool) [reply E3_Distill_ESRGAN/libs/train.py]: PS: manual_dir and data_dir are parameters of `tfds.load(...)`.
